src/dataregistry/query.py

```diff
@@ -67,149 +67,169 @@ def __init__(self, db_engine, dialect, schema_version=SCHEMA_VERSION):
         '''
-        Get specified properties for datasets satisfying all filters
+        Get specified properties for datasets satisfying all filters.
+
+        If property_names is None, return all properties from the dataset table
+        (only). Otherwise, return the property_names columns for each
+        discovered dataset (which can be from multiple tables via a join).
 
-        If property_names is None, return all properties. Otherwise it should be
-        a sublist of the return from list_dataset_properties
+        Filters should be a list of Filter objects, which are constraints on
+        column values.
 
-        filters should be a list of Filter objects, which are constraints on
-        column values
+        These choices get translated into an SQL query.
 
-        Return (probably) pandas dataframe
+        Example queries
+        ---------------
+        find_datasets(property_names=None, filters=[])
+            - SELECT * FROM registry_dev.dataset
+
+        find_datasets(property_names=["dataset.name", "execution.name"], filters=[])
+            - SELECT dataset.name, execution.name FROM registry_dev.dataset
+              JOIN registry_dev.execution ON
+              registry_dev.execution.execution_id =
+              registry_dev.dataset.execution_id
+
+        f = Filter("dataset.name", "==", "DESC dataset 1")
+        find_datasets(property_names=["dataset.description"], filters=[f])
+            - SELECT dataset.description FROM registry_dev.dataset WHERE
+              registry_dev.dataset.name = :name_1
+
+        Returns
+        -------
+        result : sqlAlchemy Result object
         '''
-        if not self._all_dataset_properties:
-            self.list_dataset_properties()
-
-        # Determine if we need a join or not
-        j = self._tables['dataset'].join(self._tables['execution'])
-        if not property_names:
-            # join dataset and execution tables
-            stmt = select('dataset_table').select_from(j)
+
+        # What tables are required for this query?
+        tables_required, _, _ = self._parse_selected_columns(property_names)
+
+        # Construct query
+
+        # No properties requested, return all from dataset table (only)
+        if property_names is None:
+            stmt = select("*").select_from(self._tables['dataset'])
+
+        # Return the selected properties.
         else:
             stmt = select(*[text(p) for p in property_names])
-            tbls = set()
-            for p in property_names:
-                tbls.update({p.split('.')[0]})
-            if len(tbls) == 2:
+
+            # Create joins
+            if len(tables_required) > 1:
+                j = self._tables["dataset"]
+                for i in range(len(tables_required)):
+                    if tables_required[i] == "dataset":
+                        continue
+
+                    j = j.join(self._tables[tables_required[i]])
```

Comment:
This might not work properly for the dependency table;  an onclause argument might be required. The dependency table now (in my `dependency` branch) links input datasets to an execution.   If someone asks for dependencies the SQL will have to look like

`select from dataset join dependency on execution_id  `

Maybe that's what sqlalchemy will do by default but it needs to be tried.

Reply:
ok, lets wait till your branch is merged into main and then we can test this branch further after those changes are applied.